packages/queued-request-controller/src/QueuedRequestMiddleware.ts

```diff
@@ -0,0 +1,168 @@
+import type {
+  AddApprovalRequest,
+  ApprovalRequest,
+} from '@metamask/approval-controller';
+import type { ControllerMessenger } from '@metamask/base-controller';
+import type { InfuraNetworkType } from '@metamask/controller-utils';
+import {
+  ApprovalType,
+  BUILT_IN_NETWORKS,
+  isNetworkType,
+} from '@metamask/controller-utils';
+import type { JsonRpcMiddleware } from '@metamask/json-rpc-engine';
+import { createAsyncMiddleware } from '@metamask/json-rpc-engine';
+import type {
+  NetworkClientId,
+  NetworkControllerFindNetworkClientIdByChainIdAction,
+  NetworkControllerGetNetworkClientByIdAction,
+  NetworkControllerGetStateAction,
+  NetworkControllerSetActiveNetworkAction,
+  NetworkControllerSetProviderTypeAction,
+} from '@metamask/network-controller';
+import { serializeError } from '@metamask/rpc-errors';
+import type { SelectedNetworkControllerSetNetworkClientIdForDomainAction } from '@metamask/selected-network-controller';
+import { SelectedNetworkControllerActionTypes } from '@metamask/selected-network-controller';
+import type { Json, JsonRpcParams, JsonRpcRequest } from '@metamask/utils';
+
+import type { QueuedRequestControllerEnqueueRequestAction } from './QueuedRequestController';
+import { QueuedRequestControllerActionTypes } from './QueuedRequestController';
+
+const isConfirmationMethod = (method: string) => {
+  const confirmationMethods = [
+    'eth_sendTransaction',
+    'wallet_watchAsset',
+    'wallet_switchEthereumChain',
+    'eth_signTypedData_v4',
+  ];
+
+  return !confirmationMethods.includes(method);
+};
+
+export const createQueuedRequestMiddleware = (
+  messenger: ControllerMessenger<
+    | QueuedRequestControllerEnqueueRequestAction
+    | NetworkControllerGetStateAction
+    | NetworkControllerSetActiveNetworkAction
+    | NetworkControllerSetProviderTypeAction
+    | NetworkControllerGetNetworkClientByIdAction
+    | NetworkControllerFindNetworkClientIdByChainIdAction
+    | SelectedNetworkControllerSetNetworkClientIdForDomainAction
+    | AddApprovalRequest,
+    never
+  >,
+  useRequestQueue: () => boolean,
+): JsonRpcMiddleware<JsonRpcParams, Json> => {
+  return createAsyncMiddleware(
+    async (
+      req: JsonRpcRequest & {
+        origin?: string;
+        networkClientId?: NetworkClientId;
+      },
+      res,
+      next,
+    ) => {
+      const { origin, networkClientId } = req;
+
+      if (!origin) {
+        throw new Error("Request object is lacking an 'origin'");
+      }
+
+      if (!networkClientId) {
+        throw new Error("Request object is lacking a 'networkClientId'");
+      }
+
+      if (!useRequestQueue() || isConfirmationMethod(req.method)) {
+        next();
+        return;
+      }
+
+      const networkClientIdForRequest = req.networkClientId as NetworkClientId;
```

Comment:
The `as` casting is unnecessary now that we have proper type narrowing.
```suggestion
      const networkClientIdForRequest = networkClientId;
```

Reply:
can actually get rid of this now duplicated value! Ty :)